README.md
# NAME

App::cpanminus - get, unpack, build and install modules from CPAN

# SYNOPSIS

    cpanm Module

Run `cpanm -h` or `perldoc cpanm` for more options.

# DESCRIPTION

cpanminus is a script to get, unpack, build and install modules from
CPAN and does nothing else.

It's dependency free (can bootstrap itself), requires zero
configuration, and stands alone. When running, it requires only 10MB
of RAM.

# INSTALLATION

There are several ways to install cpanminus to your system.

## Package management system

There are Debian packages, RPMs, FreeBSD ports, and packages for other
operation systems available. If you want to use the package management system,
search for cpanminus and use the appropriate command to install. This makes it
easy to install `cpanm` to your system without thinking about where to
install, and later upgrade.

## Installing to system perl

You can also use the latest cpanminus to install cpanminus itself:

    curl -L https://cpanmin.us | perl - --sudo App::cpanminus

This will install `cpanm` to your bin directory like
`/usr/local/bin` and you'll need the `--sudo` option to write to
the directory, unless you configured `INSTALL_BASE` with [local::lib](https://metacpan.org/pod/local::lib).

## Installing to local perl (perlbrew, plenv etc.)

If you have perl in your home directory, which is the case if you use
tools like [perlbrew](https://metacpan.org/pod/perlbrew) or plenv, you don't need the `--sudo` option, since
you're most likely to have a write permission to the perl's library
path. You can just do:

    curl -L https://cpanmin.us | perl - App::cpanminus

to install the `cpanm` executable to the perl's bin path, like
`~/perl5/perlbrew/bin/cpanm`.

## Downloading the standalone executable

You can also copy the standalone executable to whatever location you'd like.

    cd ~/bin
    curl -L https://cpanmin.us/ -o cpanm
    chmod +x cpanm

This just works, but be sure to grab the new version manually when you
upgrade because `--self-upgrade` might not work with this installation setup.

## Troubleshoot: HTTPS warnings

When you run `curl` commands above, you may encounter SSL handshake
errors or certification warnings. This is due to your HTTP client
(curl) being old, or SSL certificates installed on your system needs
to be updated.

You're recommended to update the software or system if you can. If
that is impossible or difficult, use the `-k` option with curl or an
alternative URL, `https://git.io/cpanm`

# DEPENDENCIES

perl 5.8.1 or later.

- 'tar' executable (bsdtar or GNU tar version 1.22 are recommended) or Archive::Tar to unpack files.
- C compiler, if you want to build XS modules.
- make
- Module::Build (core in 5.10)

# QUESTIONS

## How does cpanm get/parse/update the CPAN index?

It queries the CPAN Meta DB site at [http://cpanmetadb.plackperl.org/](http://cpanmetadb.plackperl.org/).
The site is updated at least every hour to reflect the latest changes
from fast syncing mirrors. The script then also falls back to query the
module at [http://metacpan.org/](http://metacpan.org/) using its seach API.

Upon calling these API hosts, cpanm (1.6004 or later) will send the
local perl versions to the server in User-Agent string by default. You
can turn it off with `--no-report-perl-version` option. Read more
about the option with [cpanm](https://metacpan.org/pod/cpanm), and read more about the privacy policy
about this data collection at [http://cpanmetadb.plackperl.org/#privacy](http://cpanmetadb.plackperl.org/#privacy)

Fetched files are unpacked in `~/.cpanm` and automatically cleaned up
periodically.  You can configure the location of this with the
`PERL_CPANM_HOME` environment variable.

## Where does this install modules to? Do I need root access?

It installs to wherever ExtUtils::MakeMaker and Module::Build are
configured to (via `PERL_MM_OPT` and `PERL_MB_OPT`).

By default, it installs to the site\_perl directory that belongs to
your perl. You can see the locations for that by running `perl -V`
and it will be likely something under `/opt/local/perl/...` if you're
using system perl, or under your home directory if you have built perl
yourself using perlbrew or plenv.

If you've already configured local::lib on your shell, cpanm respects
that settings and modules will be installed to your local perl5
directory.

At a boot time, cpanminus checks whether you have already configured
local::lib, or have a permission to install modules to the site\_perl
directory.  If neither, i.e. you're using system perl and do not run
cpanm as a root, it automatically sets up local::lib compatible
installation path in a `perl5` directory under your home
directory.

To avoid this, run `cpanm` either as a root user, with `--sudo`
option, or with `--local-lib` option.

## cpanminus can't install the module XYZ. Is it a bug?

It is more likely a problem with the distribution itself. cpanminus
doesn't support or may have issues with distributions such as follows:

- Tests that require input from STDIN.
- Build.PL or Makefile.PL that prompts for input even when
`PERL_MM_USE_DEFAULT` is enabled.
- Modules that have invalid numeric values as VERSION (such as `1.1a`)

These failures can be reported back to the author of the module so
that they can fix it accordingly, rather than to cpanminus.

## Does cpanm support the feature XYZ of [CPAN](https://metacpan.org/pod/CPAN) and [CPANPLUS](https://metacpan.org/pod/CPANPLUS)?

Most likely not. Here are the things that cpanm doesn't do by
itself.

If you need these features, use [CPAN](https://metacpan.org/pod/CPAN), [CPANPLUS](https://metacpan.org/pod/CPANPLUS) or the standalone
tools that are mentioned.

- CPAN testers reporting. See [App::cpanminus::reporter](https://metacpan.org/pod/App::cpanminus::reporter)
- Building RPM packages from CPAN modules
- Listing the outdated modules that needs upgrading. See [App::cpanoutdated](https://metacpan.org/pod/App::cpanoutdated)
- Showing the changes of the modules you're about to upgrade. See [cpan-listchanges](https://metacpan.org/pod/cpan-listchanges)
- Patching CPAN modules with distroprefs.

See [cpanm](https://metacpan.org/pod/cpanm) or `cpanm -h` to see what cpanminus _can_ do :)

# COPYRIGHT

Copyright 2010- Tatsuhiko Miyagawa

The standalone executable contains the following modules embedded.

- [CPAN::DistnameInfo](https://metacpan.org/pod/CPAN::DistnameInfo) Copyright 2003 Graham Barr
- [local::lib](https://metacpan.org/pod/local::lib) Copyright 2007-2009 Matt S Trout
- [HTTP::Tiny](https://metacpan.org/pod/HTTP::Tiny) Copyright 2011 Christian Hansen
- [Module::Metadata](https://metacpan.org/pod/Module::Metadata) Copyright 2001-2006 Ken Williams. 2010 Matt S Trout
- [version](https://metacpan.org/pod/version) Copyright 2004-2010 John Peacock
- [JSON::PP](https://metacpan.org/pod/JSON::PP) Copyright 2007-2011 by Makamaka Hannyaharamitu
- [CPAN::Meta](https://metacpan.org/pod/CPAN::Meta), [CPAN::Meta::Requirements](https://metacpan.org/pod/CPAN::Meta::Requirements) Copyright (c) 2010 by David Golden and Ricardo Signes
- [CPAN::Meta::YAML](https://metacpan.org/pod/CPAN::Meta::YAML) Copyright 2010 Adam Kennedy
- [File::pushd](https://metacpan.org/pod/File::pushd) Copyright 2012 David Golden

# LICENSE

This software is licensed under the same terms as Perl.

# CREDITS

## CONTRIBUTORS

Patches and code improvements were contributed by:

Goro Fuji, Kazuhiro Osawa, Tokuhiro Matsuno, Kenichi Ishigaki, Ian
Wells, Pedro Melo, Masayoshi Sekimura, Matt S Trout (mst), squeeky,
horus and Ingy dot Net.

## ACKNOWLEDGEMENTS

Bug reports, suggestions and feedbacks were sent by, or general
acknowledgement goes to:

Jesse Vincent, David Golden, Andreas Koenig, Jos Boumans, Chris
Williams, Adam Kennedy, Audrey Tang, J. Shirley, Chris Prather, Jesse
Luehrs, Marcus Ramberg, Shawn M Moore, chocolateboy, Chirs Nehren,
Jonathan Rockway, Leon Brocard, Simon Elliott, Ricardo Signes, AEvar
Arnfjord Bjarmason, Eric Wilhelm, Florian Ragwitz and xaicron.

# COMMUNITY

- [http://github.com/miyagawa/cpanminus](http://github.com/miyagawa/cpanminus) - source code repository, issue tracker
- [irc://irc.perl.org/#cpanm](irc://irc.perl.org/#cpanm) - discussions about cpanm and its related tools

# NO WARRANTY

This software is provided "as-is," without any express or implied
warranty. In no event shall the author be held liable for any damages
arising from the use of the software.

# SEE ALSO

[CPAN](https://metacpan.org/pod/CPAN) [CPANPLUS](https://metacpan.org/pod/CPANPLUS) [pip](https://metacpan.org/pod/pip)


Changes
See http://github.com/miyagawa/cpanminus/ for the latest development.

{{$NEXT}}

1.7040 2015-10-02T08:33:50Z
 1.7040  2015-10-02 09:43:30 CET
   [Improvements]
      - recurse the submodules or update them on cloning and checkout
 
1.7039  2015-06-28 18:05:10 PDT
   [Improvements]
      - Only enable softening of MakeMaker dependency when Module::Install is present (reported by haarg)
      - Support .git URL with http protocol (shoichikaji) #468

1.7038  2015-06-22 18:04:44 PDT
   [Releng]
      - bump minimum perl requirement to 5.8.1 in META files

1.7037  2015-06-18 14:37:32 PDT
   [Improvements]
      - Bump CPAN-Meta-Check to 0.012 #465
      - Bump HTTP-Tiny to 0.056
      - Bump CPAN-Meta to 2.150005
      - Bump CPAN-Meta-YAML to 0.016

1.7036  2015-06-05 22:07:50 PDT
   [Improvements]
      - Also soften MakeMaker depenency in runtime prereq, which is just a weird bug in an old
        Module::Install pre 1.0

1.7035  2015-06-05 10:51:17 PDT
   [Improvements]
      - Add a workaround for distributions with an old (buggy) Module::Install, now that the problems
        occur more frequently with 5.22.0 shipping with a dev version of ExtUtils::MakeMaker #463

1.7034  2015-05-07 14:20:57 PDT
   [Improvements]
      - Improved an error message when dependency requirement in cpanfile get conflict with prereqs
        in sub dependencies.

1.7033  2015-05-01 17:17:42 PDT
   [Improvements]
      - Remove MakeMaker and Module::Build from cpanm's runtime dependencies. Rather it will upgrade
        the minimum version requirement as it finds distributions that uses them in configure
        requirements. Also, they will be installed with 'notest' option for now, since Module::Build
        test suite takes a bit too long time.

1.7032  2015-04-29 18:51:36 PDT
   [Improvements]
      - Prefer to use cpanmetadb's package history API for pinning versions. Note that it only
        works against versions since April 2012. You can set `--metacpan` to still prefer MetaCPAN
        API search, and can use `--cascade-search` to fallback to MetaCPAN after CPAN MetaDB as well.

1.7031  2015-04-22 14:13:37 PDT
   [Bug Fixes]
      - Fixed a rare case where circular dependencies in cpanfile prevents satisfying modules from
        being installed. (#452)

   [Improvements]
      - Stop reporting perl versions by default if running under certain CI related env vars.

1.7030  2015-04-19 14:15:34 CEST
      - Same as 1.7029

1.7029  2015-04-19 00:15:32 CEST
   [New Features]
      - Consider root cpanfile's requirement when following dependencies with --installdeps (#363)
      - Added back --metacpan option

   [Improvements]
      - Bump Module::Metadata to 1.000027
      - Bump CPAN::Meta::Check to 0.011
      - Bump CPAN::Meta::YAML to 0.014
      - Bump Parse::PMFile to 0.36

1.7028  2015-04-17 19:22:59 CEST
   [Improvements]
      - Keep the mirror index file as absolute (skaji) #438
      - Updated the documentation about HTTPS cert warnings #441

1.7027  2015-02-14 12:14:13 PST
   [Improvements]
      - Use provides as provided in META.json and do not scan packages as PAUSE does. #435

1.7026  2015-02-13 17:11:38 PST
   [Improvements]
      - Respect provides in META file before loading from Parse::PMFile

1.7025  2015-02-07 07:58:21 CET
   [Improvements]
      - Bump Parse::PMFile to 0.35
      - Bump CPAN::Meta::Check to 0.020
      - Bump CPAN::Meta::Requirements to 2.132
      - Bump HTTP::Tiny to 0.054
      - Bump Module::Metadata to 1.000026
      - Bump version.pm to 0.9912

1.7024  2015-01-12 13:31:19 PST
   [New Features]
      - Includes vendor lib and arch path to search for core modules, when installing self-contained
        lib by default. This behavior can be turned off with --exclude-vendor (tartansandal) #426

1.7023  2015-01-05 07:59:52 JST
   [Improvements]
      - Bump Parse::PMFile to 0.34
      - Bump CPAN::Meta::Requirements to 2.131
      - Bump HTTP::Tiny to 0.53
      - Bump Module::Metadata to 1.000025
      - Bump local::lib to 2.000015

1.7022  2014-12-12 16:41:47 PST
   [Improvements]
      - Enabled unsafe option to parse version with PMFile to avoid segfaults and weird errors

1.7021  2014-12-11 21:42:36 PST
   [Improvements]
      - Bump Parse::PMFile to 0.32
      - Bump HTTP::Tiny to 0.052

1.7020  2014-12-08 17:53:18 PST
   [Improvements]
      - Bump Parse::PMFile to 0.31

1.7019  2014-12-04 12:51:41 PST
   [Improvements]
      - Bump Parse::PMFile to 0.30

1.7018  2014-11-25 14:08:41 PST
   [Maintenance]
      - bump $VERSION correctly in cpanm executable

1.7017  2014-11-25 14:00:40 PST
   [Improvements]
      - Bump fatpacked CPAN::Meta and HTTP::Tiny
      - Added documentation for -M and --from in man page

1.7016  2014-11-16 11:46:32 PST
   [New Features]
      - Added -M and --from to set --mirror and --mirror-only at the same time while
        overriding previously set mirrors (sri) #175, #417

   [Improvements]
      - cpanmin.us is now available under HTTPS (marcus)

1.7015  2014-11-14 13:14:07 PST
   [Improvements]
      - Bump Parse::PMFile to 0.29
      - Bump CPAN::Meta::Requirements to 2.129

1.7014  2014-10-07 20:52:58 PDT
   [Improvements]
      - Bump Parse::PMFile to 0.28

1.7013  2014-10-06 23:52:14 PDT
   [Improvements]
      - Update bundled JSON::PP to fix issues with JSON parsing on older perl 5.8

1.7012  2014-09-26 19:28:25 PDT
   [Improvements]
      - Upgrade CPAN::Meta and HTTP::Tiny in fatpack
      - Skip @INC on dev directory when running from git checkout when to search
        for an installed module

1.7011  2014-09-22 15:08:22 JST
      - Same as 1.7010. Make it a non-trial release.

1.7010  2014-09-17 18:27:53 JST
   [Improvements]
      - Use Parse::PMFile rather than our own fork (charsbar) #409

1.7009  2014-09-10 01:43:45 PDT
   [Bug Fixes]
      - Fix fatpacked executable to include Parse::CPAN::Meta

1.7008  2014-09-10 01:19:04 PDT
   [Improvements]
      - Show **** instead of *password* when masking passwords

lib/App/cpanminus.pm
package App::cpanminus;
our $VERSION = "1.7040";

=encoding utf8

=head1 NAME

App::cpanminus - get, unpack, build and install modules from CPAN

=head1 SYNOPSIS

    cpanm Module

Run C<cpanm -h> or C<perldoc cpanm> for more options.

=head1 DESCRIPTION

cpanminus is a script to get, unpack, build and install modules from
CPAN and does nothing else.

It's dependency free (can bootstrap itself), requires zero
configuration, and stands alone. When running, it requires only 10MB
of RAM.

=head1 INSTALLATION

There are several ways to install cpanminus to your system.

=head2 Package management system

There are Debian packages, RPMs, FreeBSD ports, and packages for other
operation systems available. If you want to use the package management system,
search for cpanminus and use the appropriate command to install. This makes it
easy to install C<cpanm> to your system without thinking about where to
install, and later upgrade.

=head2 Installing to system perl

You can also use the latest cpanminus to install cpanminus itself:

    curl -L https://cpanmin.us | perl - --sudo App::cpanminus

This will install C<cpanm> to your bin directory like
C</usr/local/bin> and you'll need the C<--sudo> option to write to
the directory, unless you configured C<INSTALL_BASE> with L<local::lib>.

=head2 Installing to local perl (perlbrew, plenv etc.)

If you have perl in your home directory, which is the case if you use
tools like L<perlbrew> or plenv, you don't need the C<--sudo> option, since
you're most likely to have a write permission to the perl's library
path. You can just do:

    curl -L https://cpanmin.us | perl - App::cpanminus

to install the C<cpanm> executable to the perl's bin path, like
C<~/perl5/perlbrew/bin/cpanm>.

=head2 Downloading the standalone executable

You can also copy the standalone executable to whatever location you'd like.

    cd ~/bin
    curl -L https://cpanmin.us/ -o cpanm
    chmod +x cpanm

This just works, but be sure to grab the new version manually when you
upgrade because C<--self-upgrade> might not work with this installation setup.

=head2 Troubleshoot: HTTPS warnings

When you run C<curl> commands above, you may encounter SSL handshake
errors or certification warnings. This is due to your HTTP client
(curl) being old, or SSL certificates installed on your system needs
to be updated.

You're recommended to update the software or system if you can. If
that is impossible or difficult, use the C<-k> option with curl or an
alternative URL, C<https://git.io/cpanm>

=head1 DEPENDENCIES

perl 5.8.1 or later.

=over 4

=item *

'tar' executable (bsdtar or GNU tar version 1.22 are recommended) or Archive::Tar to unpack files.

=item *

C compiler, if you want to build XS modules.

=item *

make

=item *

Module::Build (core in 5.10)

=back

=head1 QUESTIONS

=head2 How does cpanm get/parse/update the CPAN index?

It queries the CPAN Meta DB site at L<http://cpanmetadb.plackperl.org/>.
The site is updated at least every hour to reflect the latest changes
from fast syncing mirrors. The script then also falls back to query the
module at L<http://metacpan.org/> using its seach API.

Upon calling these API hosts, cpanm (1.6004 or later) will send the
local perl versions to the server in User-Agent string by default. You
can turn it off with C<--no-report-perl-version> option. Read more
about the option with L<cpanm>, and read more about the privacy policy
about this data collection at L<http://cpanmetadb.plackperl.org/#privacy>

Fetched files are unpacked in C<~/.cpanm> and automatically cleaned up
periodically.  You can configure the location of this with the
C<PERL_CPANM_HOME> environment variable.

=head2 Where does this install modules to? Do I need root access?

It installs to wherever ExtUtils::MakeMaker and Module::Build are
configured to (via C<PERL_MM_OPT> and C<PERL_MB_OPT>).

By default, it installs to the site_perl directory that belongs to
your perl. You can see the locations for that by running C<perl -V>
and it will be likely something under C</opt/local/perl/...> if you're
using system perl, or under your home directory if you have built perl
yourself using perlbrew or plenv.

If you've already configured local::lib on your shell, cpanm respects
that settings and modules will be installed to your local perl5
directory.

At a boot time, cpanminus checks whether you have already configured
local::lib, or have a permission to install modules to the site_perl
directory.  If neither, i.e. you're using system perl and do not run
cpanm as a root, it automatically sets up local::lib compatible
installation path in a C<perl5> directory under your home
directory.

To avoid this, run C<cpanm> either as a root user, with C<--sudo>
option, or with C<--local-lib> option.

=head2 cpanminus can't install the module XYZ. Is it a bug?

It is more likely a problem with the distribution itself. cpanminus
doesn't support or may have issues with distributions such as follows:

=over 4

=item *

Tests that require input from STDIN.

=item *

Build.PL or Makefile.PL that prompts for input even when
C<PERL_MM_USE_DEFAULT> is enabled.

=item *

Modules that have invalid numeric values as VERSION (such as C<1.1a>)

=back

These failures can be reported back to the author of the module so
that they can fix it accordingly, rather than to cpanminus.

=head2 Does cpanm support the feature XYZ of L<CPAN> and L<CPANPLUS>?

Most likely not. Here are the things that cpanm doesn't do by
itself.

If you need these features, use L<CPAN>, L<CPANPLUS> or the standalone
tools that are mentioned.

=over 4

=item *

CPAN testers reporting. See L<App::cpanminus::reporter>

=item *

Building RPM packages from CPAN modules

=item *

Listing the outdated modules that needs upgrading. See L<App::cpanoutdated>

=item *

Showing the changes of the modules you're about to upgrade. See L<cpan-listchanges>

=item *

Patching CPAN modules with distroprefs.

=back

See L<cpanm> or C<cpanm -h> to see what cpanminus I<can> do :)

=head1 COPYRIGHT

Copyright 2010- Tatsuhiko Miyagawa

The standalone executable contains the following modules embedded.

=over 4

=item L<CPAN::DistnameInfo> Copyright 2003 Graham Barr

=item L<local::lib> Copyright 2007-2009 Matt S Trout

=item L<HTTP::Tiny> Copyright 2011 Christian Hansen

=item L<Module::Metadata> Copyright 2001-2006 Ken Williams. 2010 Matt S Trout

=item L<version> Copyright 2004-2010 John Peacock

=item L<JSON::PP> Copyright 2007-2011 by Makamaka Hannyaharamitu

=item L<CPAN::Meta>, L<CPAN::Meta::Requirements> Copyright (c) 2010 by David Golden and Ricardo Signes

=item L<CPAN::Meta::YAML> Copyright 2010 Adam Kennedy

=item L<File::pushd> Copyright 2012 David Golden

=back

=head1 LICENSE

This software is licensed under the same terms as Perl.

=head1 CREDITS

=head2 CONTRIBUTORS

Patches and code improvements were contributed by:

Goro Fuji, Kazuhiro Osawa, Tokuhiro Matsuno, Kenichi Ishigaki, Ian
Wells, Pedro Melo, Masayoshi Sekimura, Matt S Trout (mst), squeeky,
horus and Ingy dot Net.

=head2 ACKNOWLEDGEMENTS

Bug reports, suggestions and feedbacks were sent by, or general
acknowledgement goes to:

Jesse Vincent, David Golden, Andreas Koenig, Jos Boumans, Chris
Williams, Adam Kennedy, Audrey Tang, J. Shirley, Chris Prather, Jesse
Luehrs, Marcus Ramberg, Shawn M Moore, chocolateboy, Chirs Nehren,
Jonathan Rockway, Leon Brocard, Simon Elliott, Ricardo Signes, AEvar
Arnfjord Bjarmason, Eric Wilhelm, Florian Ragwitz and xaicron.

=head1 COMMUNITY

=over 4

=item L<http://github.com/miyagawa/cpanminus> - source code repository, issue tracker

=item L<irc://irc.perl.org/#cpanm> - discussions about cpanm and its related tools

=back

=head1 NO WARRANTY

This software is provided "as-is," without any express or implied
warranty. In no event shall the author be held liable for any damages
arising from the use of the software.

=head1 SEE ALSO

L<CPAN> L<CPANPLUS> L<pip>

=cut

1;
